stores/bookstore.ts

```diff
@@ -93,6 +107,89 @@ export const useBookstoreStore = defineStore('bookstore', () => {
     }
   }
 
+  /* Bookstore Search Results */
+
+  const bookstoreSearchResultsByQueryMap = ref<Record<string, BookstoreSearchResults>>({})
+
+  const getBookstoreSearchResultsByQuery = computed(() => (query: string) => {
+    const items
+      = (bookstoreSearchResultsByQueryMap.value[query]?.items || [])
+        .filter(item => !!getBookstoreInfoByNFTClassId.value(item.classId))
+    return {
+      items,
+      isFetchingItems: bookstoreSearchResultsByQueryMap.value[query]?.isFetching || false,
+      hasFetchedItems: bookstoreSearchResultsByQueryMap.value[query]?.hasFetched || false,
+      nextItemsKey: bookstoreSearchResultsByQueryMap.value[query]?.nextKey || undefined,
+    }
+  })
+
+  async function fetchSearchResults(type: 'author' | 'publisher', searchTerm: string, {
+    isRefresh = false,
+  }: {
+    isRefresh?: boolean
+  } = {}) {
+    const { fetchNFTClassesByAuthorName, fetchNFTClassesByPublisherName } = await import('~/shared/utils/indexer')
+
+    const queryKey = `${type}:${searchTerm}`
+
+    if (bookstoreSearchResultsByQueryMap.value[queryKey]?.isFetching) {
+      return
+    }
+
+    if (!bookstoreSearchResultsByQueryMap.value[queryKey] || isRefresh) {
+      bookstoreSearchResultsByQueryMap.value[queryKey] = {
+        items: [],
+        isFetching: false,
+        hasFetched: false,
+      }
+    }
+
+    try {
+      bookstoreSearchResultsByQueryMap.value[queryKey].isFetching = true
+
+      const options = {
+        limit: 100,
+        key: isRefresh ? undefined : bookstoreSearchResultsByQueryMap.value[queryKey]?.nextKey,
+      }
+
+      let result
+      if (type === 'author') {
+        result = await fetchNFTClassesByAuthorName(searchTerm, options)
+      }
+      else {
+        result = await fetchNFTClassesByPublisherName(searchTerm, options)
+      }
+
+      if (result) {
+        const nftClasses = result.data
+        const mappedItems = nftClasses
+          .map(nftClass => ({
+            classId: nftClass.address.toLowerCase(),
+            title: nftClass.name || '',
+            imageUrl: nftClass.metadata?.image || '',
+            minPrice: undefined,
+          }))
+
+        nftClasses.forEach((nftClass) => {
+          nftStore.lazyFetchNFTClassAggregatedMetadataById(nftClass.address)
+        })
```

Comment:
why fetching metadata here instead of lazy loading?

Reply:
Similar problem to recommended items, indexer return everything on chain, but they might not have a listing (bookstore info), using bookstore item for lazy load would make unlisted items show up
Maybe we can filter for bookstore info undefined vs null to avoid triggering loading here, idk

Reply:
Refactored to store `null` as special value in bookstore info